bookkeeper-server/src/main/java/org/apache/bookkeeper/client/TopologyAwareEnsemblePlacementPolicy.java

```diff
@@ -790,6 +794,113 @@ public void updateBookieInfo(Map<BookieId, BookieInfo> bookieInfoMap) {
         }
     }
 
+    @Override
+    public Map<Integer, BookieId> replaceNotAdheringPlacementPolicyBookie(List<BookieId> ensemble, int writeQuorumSize,
+            int ackQuorumSize) {
+        if (CollectionUtils.isEmpty(ensemble)) {
+            return Collections.emptyMap();
+        }
+        PlacementPolicyAdherence ensembleAdheringToPlacementPolicy = isEnsembleAdheringToPlacementPolicy(ensemble,
+                writeQuorumSize, ackQuorumSize);
+        if (PlacementPolicyAdherence.FAIL != ensembleAdheringToPlacementPolicy) {
+            return Collections.emptyMap();
+        }
+        Map<BookieId, Integer> bookieIndex = new HashMap<>();
+        for (int i = 0; i < ensemble.size(); i++) {
+            bookieIndex.put(ensemble.get(i), i);
+        }
+
+        Map<BookieId, BookieNode> clone = new HashMap<>(knownBookies);
+
+        Map<String, List<BookieNode>> toPlaceGroup = new HashMap<>();
+        for (BookieId bookieId : ensemble) {
+            //If the bookieId shutdown, put it to inactive.
+            BookieNode bookieNode = clone.get(bookieId);
```

Comment:
If the bookie shutdown, it will be removed from knownBookies immediately. It belongs to `DATA_LOSS` type

Reply:
In theory, If the the fragment is DATA_LOSS, it won't invoke this method. It will repair data loss firstly.